RecoLocalTracker/SiPixelRecHits/src/PixelCPEClusterRepair.cc

```diff
@@ -0,0 +1,628 @@
+   //    We'll later make  clustMatrix[ mrow ][ mcol ]
+   int mrow=0, mcol=0;
+   for (int i=0 ; i!=theClusterParam.theCluster->size(); ++i )
+   {
+      auto pix = theClusterParam.theCluster->pixel(i);
+      int irow = int(pix.x);
+      int icol = int(pix.y);
+      mrow = std::max(mrow,irow);
+      mcol = std::max(mcol,icol);
+   }
+   mrow -= row_offset; mrow+=1; mrow = std::min(mrow,cluster_matrix_size_x);
+   mcol -= col_offset; mcol+=1; mcol = std::min(mcol,cluster_matrix_size_y);
+   assert(mrow>0); assert(mcol>0);
+
+
+   //--- Make and fill the bool arrays flagging double pixels
+   bool xdouble[mrow], ydouble[mcol];
+   // x directions (shorter), rows
+   for (int irow = 0; irow < mrow; ++irow)
+      xdouble[irow] = theDetParam.theRecTopol->isItBigPixelInX( irow+row_offset );
+   //
+   // y directions (longer), columns
+   for (int icol = 0; icol < mcol; ++icol)
+      ydouble[icol] = theDetParam.theRecTopol->isItBigPixelInY( icol+col_offset );
+
+   //--- C-style matrix.  We'll need it in either case.
+   float clustMatrix[mrow][mcol];
+   float clustMatrix2[mrow][mcol];
+
+   //--- Prepare struct that passes pointers to TemplateReco.  It doesn't own anything.
+   SiPixelTemplateReco::ClusMatrix   clusterPayload  { &clustMatrix[0][0], xdouble, ydouble, mrow,mcol};
+   SiPixelTemplateReco2D::ClusMatrix clusterPayload2d{ &clustMatrix2[0][0], xdouble, ydouble, mrow,mcol};
+
+
+   //--- Copy clust's pixels (calibrated in electrons) into clustMatrix;
+   memset( clustMatrix, 0, sizeof(float)*mrow*mcol );   // Wipe it clean.
+   for (int i=0 ; i!=theClusterParam.theCluster->size(); ++i )
+   {
+      auto pix = theClusterParam.theCluster->pixel(i);
+      int irow = int(pix.x) - row_offset;
+      int icol = int(pix.y) - col_offset;
+      // &&& Do we ever get pixels that are out of bounds ???  Need to check.
+      if ( (irow<mrow) & (icol<mcol) ) clustMatrix[irow][icol] =  float(pix.adc);
+   }
+   // fillClustMatrix( float * clustMatrix );
+
+   //--- Save a copy of clustMatrix into clustMatrix2
+   memcpy( clustMatrix2, clustMatrix, sizeof(float)*mrow*mcol);
+
+   //--- Set both return statuses, since we may call only one.
+   theClusterParam.ierr  = 0;
+   theClusterParam.ierr2 = 0;
+
+
+
+   //--- Are we on edge?
+   if ( theClusterParam.isOnEdge_ ) {
+     //--- Call the Template Reco 2d with cluster repair.0
+     callTempReco3D( theDetParam, theClusterParam, clusterPayload2d, ID, lp );
+   }
+   else {
+     theClusterParam.recommended3D_ = false;
+     //--- Call the vanilla Template Reco
+     callTempReco2D( theDetParam, theClusterParam, clusterPayload, ID, lp );
+
+     //--- Did we find a cluster which has bad probability and not enough charge?
+     if ( theClusterParam.recommended3D_ ) {
+       //--- Yes. So run Template Reco 2d with cluster repair.
+       
+       // //--- Once again (!) copy clust's pixels (calibrated in electrons) into 
```

Comment:
Are these commented out lines needed? If so, please add another line of comment which explain why; otherwise just remove them

Reply:
Same as above: I'd like to keep the commented lines for now.

Reply:
Note that there was a duplication of this snippet of code, and I bypassed it with a straight memcpy.  However, I'd like to keep this around for a little longer, just in case.